WIP: initial draft of dynamic resource allocation KEP

pohly opened the pull request:
This PR is meant for discussion around this KEP before we decide whether we actually want to start the official KEP process (create tracking issue, submit PR to k8s.io/enhancements).

Please comment here. Larger text changes can also be submitted as PRs against the `dynamic-resource-allocation` branch in this repo.

pohly committed:
WIP: initial draft of dynamic resource allocation KEP

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+objects stored in the kube-apiserver. Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
+mechanism, using a new plugin type and a new gRPC interface. Towards
+the container runtime, kubelet and those plugins then must use the
```
I don't think CDI is a must if the plugins don't deal with devices, right?

pohly replied:
Once it comes to making the resource available to a container runtime, the plugin and kubelet must use CDI. There is no other mechanism that could be used to enable the additional resources, or if there is one, no support for that is planned as part of this KEP.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discreet,
+countable per-node custom resources was added. The device plugin
```
*discrete*, how about extended resources instead of custom resources (to not mix with CRs/CRDs)?

pohly replied:
Makes sense, will change.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device.
```
we could be more specific than "doesn't allow" to give the root cause. AFAIU the limitation here is that the API does not provide enough information to be able to make decisions about partial allocations and partly because its not possible to specify desired device properties (same as above with device init).

pohly replied:
How about adding ".... because parameters are too limited and Kubernetes doesn't have enough information about the extended resources on a node to decide whether they can be shared." ?

mythi replied:
yes, sounds OK

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+  on the container image, and device plugins do not have access to the
+  container image.
+
+- *Perform runtime-specific operations*: When deploying a container
```
*Perform container runtime specific operations*

pohly replied:
Changed.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+
+   // An allocated resource is available on nodes that match this
+   // selector. If nil, the resource is available everywhere.
+   Topology *corev1.NodeSelector
```
how about `AvailableNodes` or something. `Topology` is somewhat ambiguous.

pohly replied:
Agreed, a better term would be useful. It came from CSI.

I don't like `AvailableNodes` because its not the nodes which are available. `AvailableOnNodes` is more accurate, but longer.

mythi replied:
but OTOH if CSI uses `Topology` for the exact same purpose, keeping the same field name might also be justified. my thinking was to harmonize PotentialNodes/SuitableNodes/Available(On)Nodes.

pohly replied:
The topology concept in CSI is more complex, for example it involves mapping key/value pairs provided by the driver onto Kubernetes nodes. We can and should use something simpler here - we just need a good name.

mythi replied:
+1

pohly replied:
So did we agree on `AvailableOnNodes`?

mythi replied:
`AvailableOnNodes` sounds good to me

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+   ResourceClaimName *string
+
+    // Will be used to create a stand-alone ResourceClaim to allocate the resource.
+    // The pod in which this EphemeralVolumeSource is embedded will be the
```
`PodResource`?

pohly replied:
Yes. Cut-and-paste... :grin: Fixed.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or the CO does not know if it failed or not, it
```
What's CO?

pohly replied:
Container Orchestrator.

The term is defined in the full CSI spec from which we copied. We are no longer trying to support different COs, so we can replace with Kubernetes everywhere.

In this context we can be even more specific and say "kubelet". Done.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+
+
+#### Implementing a plugin for node resources
```
this section still talks about CRDs

pohly replied:
This is intentional. It explains how a plugin could use CRDs to implement its side of the KEP.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
maybe clarify that it's `v1beta1`.

pohly replied:
I replaced "stable" with "beta".

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
ref: https://github.com/kubernetes/enhancements/pull/1949

pohly replied:
Added.

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
would it be better to say "such annotations were..."?

pohly replied:
Changed into "Passing annotations as additional parameters was".

mythi commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```
maybe we could cover the CDI spec + device plugins alternative too.

pohly replied:
@bart0sh: has that alternative been discussed? Can you add a section for it?

bart0sh replied:
@pohly I'm not sure we need that. We can't do much with device plugins without changing their API. I'v already mentioned that in this KEP.

mythi replied:
how about we mention CDI spec separately under the device plugin option still? My understanding was that the CDI spec alone (+device plugins) would help address some of the limitations mentioned under Motivation.

bart0sh replied:
That would make sense. Will do.

bart0sh replied:
@mythi @pohly done: https://github.com/pohly/enhancements/pull/3. Please review.

mythi reviewed (commented):
@pohly @bart0sh I did a first pass review. sorry that it took so long.

pohly committed:
address Mikko's review feedback

pohly commented:
I'll submit changes as separate commits because GitHub occasionally gets confused and drops comments when rewriting history. Eventually all of this will be squashed.

bart0sh committed:
update 'Device Plugins' section according to review comments

mythi commented:
Hmm. For some reason I'm not able to resolve my comments that have been addressed.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+objects stored in the kube-apiserver. Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
+mechanism, using a new plugin type and a new gRPC interface. Towards
+the container runtime, kubelet and those plugins then must use the
```
It is better to clarify how kubelet will use CDI and expose the new resource detail. E.g. will new device plugin register its capability with kubelet via the CDI? Or is it a pull model?

pohly replied:
The API between kubelet and the local plugin needs to be specified in more detail. So far we only have the gRPC calls in the KEP which describe the resource operations, but not the ones for plugin registration.

CDI itself does not cover that at all.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device.
```
We may need to address the security concern of sharing the device plugin between multiple containers.

pohly replied:
My understanding is that the plugin will handle that and if it can't, refuse to share devices.

pohly replied:
The bigger question for this KEP is around who is allowed to use these additional resources and how much of those, i.e. quota enforcement.

Given that Kubernetes doesn't have any semantic understanding of "resource quantities", it seems unlikely that this can be something that gets enforced by Kubernetes itself. A plugin might have to solve this with its own custom admission controller or after the fact (= pod and/or resource claim already created) when allocating resources.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
```
This is a good point!  I think addressing this limitation will help Kubernetes to support more use cases

pohly replied:
We need to check how well the current API supports this.

I think it can be covered by a plugin specific parameter that says "allocate up to this amount of something", but that may not be good enough. What if the requested storage class is not even defined? This is currently an error.

What if it is available, but allocation fails once? How often should Kubernetes retry before moving on without the resource?

We probably need additional parameters in ResourceClaimSpec to define the expected behavior on a case-by-case basis.

kad replied:
It is fine that allocation fails on the time of pod start time if the request for resource is optional. If we are talking about ReplicaSet of similar deployments with dynamic changing set of running pods such optional allocation failures is expected and valid behaviour.

Absence of requested class is definitely an error.

pohly replied:
> It is fine that allocation fails on the time of pod start time if the request for resource is optional.

Allocation as designed at the moment happens before the pod is allowed to start because information about where the resource was allocated is needed for choose the node for pod.

Are you suggesting that for optional resources, pod scheduling should proceed without waiting for the resource allocation and then for the assigned node, allocation is tried once with "resource not available" not blocking pod startup?

kad replied:
No, what I'm suggesting, is that in case of "optional", allocation by CDI plugin will succeed, returning topology key for any node but actual "allocation" would be "nil". meaning, CDI json might be empty - no new devices injected into OCI spec.

pohly replied:
So that's the "can be covered by a plugin specific parameter" approach that I mentioned earlier. I'm fine with that, in particular because it keeps the KEP simpler. If we need something more complex, we can still extend the KEP later.

I'll add a section explaining this approach.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (PCIe, CXL,
+  Network, special links, etc).
+
```
which layer/component (scheduler? kubelet? containerd/CRI-O?) will make the decision about which devices over the fabric are needed by the container?

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+  runtime-specific operations that may need to occur as a result of
+  injecting a device into a container (device plugins are runtime
+  agnostic). A good example is supporting GPU passthrough
+  virtualization on kata vs. runc.
```
What runtime-specific operations are needed? I do not understand what is meant by "GPU passthrough  virtualization on kata vs. runc". 
"

pohly replied:
@kad, @bart0sh: can you clarify?

mythi replied:
typically device passthrough to a VM sandbox (e.g., created by kata-containers container runtime) takes the GPU device bound to `/dev/vfio/xyx` and passes it to VMM. this device shows up as `/dev/dri/cardX` in the VM for the container to use.

In case of runc, the device on the host is `/dev/dri/cardX` which is added to the container.

if a container wants "gpu" resource, CDI needs to know if that is consumed by kata or runc to prepare the device correctly

kad replied:
@cathyhongzhang as @mythi  mentioned, device passtrough should be handled by hypervisor of VM based runtimes as example. It might be more complex cases, e.g. for FPGA to perform configuration of device via management port on host side and to VM provide access only to "user" port device node.

pohly replied:
Is that covered by CDI internally or is that something that we need to take care of in this KEP?

Right now, both allocation and the  `NodePrepareResource` call are agnostic about the runtime that will be used.

This also makes it questionable whether a resource can be used by more than one pod. What if one pod uses runc and the other kata?

kad replied:
runtime side CDI spec (json) specifies host devices (nodes, permissions, etc)... This is will become part of OCI spec that containerd/cri-o using to communicate with runc/kata/others. This KEP scope is up to creating those CDI json files. The rest is out of scope.

pohly replied:
So can we remove this particular item from the KEP?

kad replied:
no, this scenario described in this paragraph is still valid: the cdi.json file might include instructions that potentially can be runtime specific. we had discussion this week COD meeting where Evan mentioned that potentially some conditional fields are needed in CDI spec, thus this paragraph is just sub-case of conditional parameters injection.

pohly replied:
So it is relevant for CDI, but not for dynamic resource allocation, right? This KEP isn't about what the plugins put into the JSON files.

We probably should make it clearer which of these items are the motivation for CDI and which of them motivate specific design choices in this KEP. For a reader of this KEP it will be confusing to see limitations listed which then don't get addressed as part of the KEP.

pohly replied:
I have re-arranged the "motivation" section in https://github.com/pohly/enhancements/pull/4

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+
+  *Limitation*: At present, there is a chicken-and-egg problem to
+  supporting this because the device plugin is the sole dictator of
+  which containers will have these operations performed on them.
```
It is not clear to me what is meant by "all of the operations that need to occur  as part of “injecting” a device into a container, also occur for the  “plugin container” itself". Do you mean the same operations need to be applied to the plugin and the container?

pohly replied:
I copied the motivation section from an earlier document written by @kad and raised the same question myself, without getting it resolved before moving the text here.

@kad: can you clarify what was meant here?

kad replied:
@cathyhongzhang @pohly "all operations" means in that scope preparation of the device itself (e.g. FPGA programming before it can be used) or inspecting variables, container meta-data (e.g. annotations) to select right set of mounts (e.g. libraries in NVIDIA case). 

Same goes to monitoring tasks: e.g. if you have "device plugin" daemon set, for some devices to properly detect parameters or monitoring health of devices, DaemonSet should have access to control devices and be able to perform e.g. `ioctl` calls to it. So, it might be scenarios where first class of devices is used just with simple "detect over PCI ID" to be able to start real node agent that would have access to all control device nodes to properly expose device internal state/resoruces. (that example was mentioned by NVIDIA in early conversations).

elezar replied:
From the perspective of the current NVIDIA implementation, the requirements for the plugin container and an (application) container could be handled differently. For the purpose of the plugin, it is sufficient to know which devices are available and what their properties are. In an environment where configurations are not expected to change dynamically, a CDI spec (for example) could be configured once and consumed by the plugin.

Even without the CDI spec, starting the plugin container as a privileged container and including a small number of libraries as mounts would allow for the required functionality.

I think the key is whether selection by the plugin container is considered "use" of a device. The answer may be vendor / implementation specific.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+  * Arbitrary, resource-specific setup and cleanup actions
+  * Over-the-fabric resources
+  * Custom matching of resource requests with available resources
+* User-friendly API for describing resource requests
```
Do you mean the API consumed by the Kubernetes API server?

pohly replied:
We probably mean the same thing, I just would describe it the other way around: the API provided by the Kubernetes API that is used by users to describe resource requests. Kube-scheduler, kube-controller-manager and kubelet also consume that API.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
```
If no node fits, what resources must be released?

pohly replied:
Some or all of those resources that were already allocated. The exact choice will probably need a bit of experimentation to come up with a good selection for different scenarios.

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+because scheduling one Pod may affect other Pods in ways that cannot
+be anticipated by the kube-scheduler. When many Pods compete for
+limited resources, multiple attempts may be needed before a suitable
+node is found.
```
We can think about ways to mitigate the scheduling conflict.

kad replied:
@cathyhongzhang in theory yes, but that would be visible only in cases of big clusters with many device requests.
With current set of hardware and overall accelerator usages, percentage of accelerated workloads is really small, thus should be not critical for next several years, until accelerators would become more common thing.

pohly replied:
@elezar: do you agree? Is the approach proposed in this KEP going to be useful for how NVIDIA hardware is used in clouds?

cathyhongzhang commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1368 @@
+#### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
```
So this plugin interfaces with kubelet directly and generates a CDI format file. Then the high level runtime, e.g. containerd/CRI-O, will consume the CDI file and translate it to OCI format which is processed by low level runtime, such as kata container, right? If so, it means the high level runtime needs to be modified to understand the CDI format, right?

pohly replied:
Correct.

That's where the CDI dependency of this KEP comes from.

pohly commented:
> Hmm. For some reason I'm not able to resolve my comments that have been addressed.

Does it work when you go to the commit view?

https://github.com/pohly/enhancements/pull/1/commits/f118438ba88e5699c194ba355fa1dd5157b67f7f

But I see the "Resolve conversation" button also on the main page, so that's probably not it. Perhaps add a "thumbs up" reaction for all of my comments that resolve the conversation and I'll close them.

pohly committed:
Merge pull request #3 from bart0sh/PR0006-dra-add-cdi

update 'Device Plugins' section according to review comments

pohly committed:
address Mikko's review feedback

pohly committed:
Topology -> AvailableOnNodes

kad reviewed (commented):
Looks good in my opinion to start broader discussion. Added few clarifications/answers to questions.

elezar commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1372 @@
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+
```
When looking at the current device plugin API, being able to add hooks (e.g. preContainerStart) in addition to devices, mounts, and envvars would allow us to run dynamic initialisation and checks in the container. One example in the case of the current NVIDIA implementation is that we run ldconfig to update the ld.cache in the container and have the injected libraries visible to the application container. These hooks could also include checks for compatible library versions as defined by an image vendor.

pohly replied:
This leads back to the overall question: should we pursue this KEP or focus instead on incremental improvements of the existing device plugin API?

@elezar: any thoughts on this? Do you have use cases now or in the future that simply can't be handled unless we do something completely new, like this KEP?

klueska replied:
In general, incremental improvements have a higher likelihood of being accepted. It doesn't mean you can't eventually end up in the same final state that you want, it just means that the path to get there might not be as "big-bang" as originally proposed. I've seen lots of KEPs that propose large changes just fizzle out and die because its too much too fast (even if the underlying ideas were actually pretty good). Funnily enough, it will often move faster if you propose a bunch of incremental changes because it will be more obvious to the community that each incremental change is useful in its own right.

pohly replied:
As discussed in our meeting, there's no incremental path because the API changes are a core part of where we are trying to get. Without those there isn't much that we can do. Also, we are not trying to morph the existing device API into something new. Both approaches will exist alongside each other.

pohly committed:
component diagram, clarifications

pohly commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1372 @@
+
+message NodePrepareResourceResponse {
+  // The map <device name>:<kind> for all allocated devices
+ // where device name is a “name of the device” and kind is a
```
@bart0sh, @kad: can you clarify for me what the "name of the device" is?

Is that some arbitrary string made up by the node plugin? How does it relate to the CDI JSON file?

pohly replied:
Why is this a map?

pohly replied:
Is the name required to be unique across all JSON files?

For example, suppose I have two JSON files:
```
{ "kind": "foo", "devices": [ { "name": "something" } ] }
```

```
{ "kind": "bar", "devices": [ { "name": "something" } ] }
```

Then with the current API, it is not possible to select both devices at the same time.

pohly replied:
Resolved via https://github.com/pohly/enhancements/pull/5

eero-t commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1372 @@
+
+  *Limitation*: Currently, there is no way to specify the equivalent
+  of a “vendor.com/device: all” selector with these semantics
+
```
One more use-case is separate request and limit values.   While extended k8s resources could already be used to specify requests to devices, k8s requires extended resource request to match limit value.  Example: GPU scheduling request and limit values (say 250 & 500 millicores).

pohly replied:
This is covered by the current proposal, isn't it?

eero-t replied:
Implicitly yes, but I though it could be explicitly spelled out as one of the use-cases.

pohly replied:
Can you propose wording for this that I can put directly in the KEP?

pohly committed:
Merge pull request #4 from pohly/diagrams

component diagram, clarifications

pohly committed:
support reuse of the same device name

vendor.com/foo and acme.org/foo might be two valid devices. Previously this
wasn't supported because "foo" as device name mapped to only one kind.

pohly committed:
Merge pull request #5 from pohly/cdi-parameters

CDI device parameters

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device because parameters are too limited
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
```
I don't know if I agree with this one. The plugin can divide up resources however it wants (assuming the the resources it manages are "divisible") and advertise them in smaller chunks. Maybe you're talking about a more dynamic splitting of resources after the initial set of resources have been advertised to the kubelet?

pohly replied:
"doesn't allow" is indeed too strong. But the current model is very basic (resources are a certain amount of individual items and can be assigned to pods in arbitrary subsets) and does not fit all cases (for example, in a two-GPU scenario, memory from both cards cannot be combined arbitrarily).

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
+  the node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
```
I can see the usefulness in this, but it will definitely require changes to the pod spec to accomplish something like this. These types of changes make it really hard for a KEP to get accepted without a very long review process. There might be ways to accomplish something like this by encoding the "optionalness" of the resource in the resource name itself, but that will never provide a general solution and will always feel somewhat hacky since it's not integrated into the K8s resource model itself.

 If we can avoid this in the first round, that would be preferable.

pohly replied:
See my other comment about needing the API changes to get started.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (PCIe, CXL,
+  Network, special links, etc).
+
+  *Limitation*: Device Plugins framework supports only local devices.
```
Assuming you only pass back /dev nodes, yes, the current plugin interface is limited in this way. We work around this with the NVIDIA GPU stack by intercepting the runtime SPEC and injecting whatever we want into it based on the value of an environment variable. I guess CDI will allow us to do something similar (for all devices types) in a more well defined way (i.e. not via a non-standard environment variable).

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+
+  *Limitation*: Device Plugins framework supports only local devices.
+
+- *Access to the container image*: When deploying a container that
+  needs access to a GPU device, I would like my container to
+  gracefully fail to start (rather than SIGSEGV at runtime) if a
+  minimum driver requirement is not met on the host.
+
+  *Limitation*: GPU driver requirements are currently stored as labels
+  on the container image, and device plugins do not have access to the
+  container image.
```
Are you proposing that the plugin have access to the image, or just have a way of doing something like this (potentially via another means than having access to the image)?

pohly replied:
@kad: can you explain what was meant with this point?

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+  * Custom matching of resource requests with available resources,
+    including handling of optional resource requests
+* User-friendly API for describing resource requests
+* Allow resource management plugins that can be developed and deployed
+  separately from Kubernetes and are independent of specific container
+  runtimes.
```
So this proposal is to design a plugin interface that is agnostic to the container orchestrator? I.e. it may or may not run as a pod in Kubernetes and potentially also on something like Docker Swarm?

pohly replied:
No. The key point here is that the plugins can developed by third-party hardware vendors and be added to a Kubernetes cluster without having to modify the core Kubernetes binaries, in particular the scheduler, or the underlying container runtime.

The plugin is still going to be specific to Kubernetes and not meant to work with other COs. I'll clarify this in the text.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+and make progress.
+-->
+
+* An extension of the model that kube-scheduler has about
+  resources. Instead, it will need information from the plugin for
```
Some of the limitations you listed above seem hard to do "correctly" without some change to the pod spec. i'm not proposing this should be a goal, but rather, I'm just suggesting that some of the "goals" may need to be adjusted.

pohly replied:
A pod spec change indeed is a goal and core part of this proposal.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+The proposal is that a plugin handles all operations that are specific
+to the resources managed by that plugin. This includes operations at
+the control plane level (tracking where in the cluster resources are
+available, helping with pod scheduling decisions, allocating resources
+when requested) as well as the node level (preparing container
```
"allocating resources when requested"

This is going to be a tough one to sell. It has a long held opinion that the kubelet itself should always have the final say in any resource allocation decisions. Its decisions can be (and even are today) influenced by the plugins, but the kubelet always makes the final decision.

pohly replied:
kubelet can have the final say about resources that it manages.

But in this KEP, kubelet is going to be oblivious of the resources managed by a plugin. All that kubelet needs to do (can can do) is reconfiguring the container runtime to expose resources.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+such a plugin, but using it will not be required and its API is not
+part of this KEP.
+
+Three new API object types get added in a new API group:
+- ResourcePlugin, not namespaced, with a description of the plugin.
+- ResourceClass, not namespaced, with privileged parameters for
+  multiple resource instances of a certain kind. All these instances
+  are provided by the same resource plugin, which is identified by a
+  field in the class.
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual meta data, a spec and a status. The
+  spec identifies the plugin that handles the resource via a class
+  name.
```
I admittedly don't feel qualified to comment on whether this is a good idea or not, but it's definitely very different than what is done today. Can someone comment on whether there is any precedence for something like this (i.e. using a CRD to hold information like this AND have an API to the kubelet for other types of operations)?

pohly replied:
This is all very similar to how volumes are handled (PersistentVolumeClaim, PersistentVolume). One improvement is that the separate PersistentVolume object gets avoided. The approach for volumes came from avoiding non-observable status in the PersistentVolumeClaim. I think there is consensus that this led to a very complex design. SIG-Storage is still working on fixing bugs with it. Tim's pending KEP where he proposes to allow status to hold non-observable information mentions volumes.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+recently [became more
+acceptable](https://github.com/kubernetes/enhancements/pull/2537). Kube-scheduler
+and plugin communicate by modifying that status. The status is also
+how Kubernetes tracks that a plugin has allocated the resource and on
+which nodes the resource is available.
```
Does this supercede the podresources API, or duplicate info from it?

pohly replied:
Neither - it complements the podresource API. For memory, CPU, and traditional extended resources the podresources API continues to be used.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+how Kubernetes tracks that a plugin has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the PV/PVC
+model for volumes because we don't need to deal with two objects when
+allocating resources and therefore don't need something like the
+volume binding controller. If desired, a resource plugin can implement
```
OK, maybe looking at how volume plugins are designed will help me clarify my comment from above better.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+Allocation of a resource happens either immediately (“immediate
+allocation”) or when a Pod needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
+claims in a new PodSpec.Resources list. Each resource in that list
```
As I mentioned before -- any changes to the PodSpec are going to be hard to get people to accept. If there are incremental changes we can propose / make without doing this, that would go a long way towards make progress on this.

sftim replied:
Could we move the changes to workload-related APIs like ReplicaSet, StatefulSet etc?
If you had to have a workload resource like a Job in order to use this resource API, could that be OK?

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+Pod as owner, and then the normal allocation of the resource takes
+place.
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. For delayed allocation, a node is selected tentatively
+and plugin(s) try to allocate their resources for that node. If that
+succeeds, the Pod can start to run. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
```
So is the proposal to (somehow) do away with the kubelet's internal device manager entirely, and have all device management happen decentralized in the plugins themselves? How would this work to coordinate with the `TopologyManager` (or some future iteration of something like this) which needs to align resources in some way (e.g. on a common NUMA node, for example)?

pohly replied:
> So is the proposal to (somehow) do away with the kubelet's internal device manager entirely

No. That continues to be used for resources which do not use the new approach.

> How would this work to coordinate with the TopologyManager (or some future iteration of something like this) which needs to align resources in some way (e.g. on a common NUMA node, for example)?

This is open for debate. My initial thought was to clone the volume manager in kubelet or add some special cases for resources to it. What kubelet needs to do is a subset of what it does for volumes: stage them, add to the container, later unstage.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+The resources allocated for a ResourceClaim can be shared by multiple
+containers in a pod. Depending on the capabilities defined in the
+ResourceClaim by the plugin, a ResourceClaim can be used exclusively
+by one pod at a time, by a certain maximum number of pods, or an
+unlimited number of pods.
```
This is certainly a nice feature -- more akin to how volumes are shared across containers in a pod. Do you plan to also support sharing across pods, or only within a pod?

pohly replied:
Both. If it is a network attached resource, then the pods do not even need to be on the same node.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+    command: ["/bin/program"]
+    resources:
+    - name: gpu_2Gb
+  resources:
+  - name: gpu_2Gb
+    template:
+      resourceClassName: "gpu"
+      parameters:
+        memory: "2Gb"
```
To be clear, it wouldn't look exactly like this, right? You still need to support the direct fields of the "traditional" `resources` key as well (i.e. request / limits and any subkeys below that, i.e. cpus, memory, (any custom resources), etc.).

pohly replied:
That's correct. It's an example that just focused on the new API, but for the sake of clarity it is worth also showing the normal resources. Will add.

pohly replied:
"resources" here is the new field at the PodSpec level. In the container, the example incorrectly also used "resources" (the existing, stable field name) instead of the new one.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Scheduling is likely to be slower when many Pods request the new
+resource types, both because scheduling such a Pod involves more
+round-trips through the API server for ResourceClaimStatus updates and
+because scheduling one Pod may affect other Pods in ways that cannot
+be anticipated by the kube-scheduler. When many Pods compete for
+limited resources, multiple attempts may be needed before a suitable
+node is found.
```
So (assuming a daemonset is used to launch a plugin), I guess there will be one entry per-node in the CRD with the resources being made available for that node by the plugin. And then the scheduler will need to parse / aggregate this information so that it can make scheduling decisions based on it? Or maybe I'M missing something about how this is supposed to work.

pohly replied:
This is explained further in the proposal section. In general I am often struggling with the KEP format: IMHO points like this (user stories, "Notes/Constraints/Caveats") are hard to present without first explaining what is being proposed.

Readers have to take these sections at face value and later come back to understand them properly.

klueska reviewed (commented):
I went through the summary, motivation, and proposal sections and left some comments. Let's discuss these offline and then revisit the document based on our discussions.

My initial inclination is to say that we should try *not* to introduce all of these changes at once, and instead look to make changes that only modify the kubelet and/or the device plugin interface. If we can scope this down to just that and then get some experience with how well it works, then we can think in the future about expanding it to modifications of the scheduler / api-server, etc.

pohly committed:
summary: clarify relation to device API

pohly committed:
fix example

It was using the wrong field name and it wasn't clear that traditional
resources get requested the same way as before.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1408 @@
+
+![components](./components.svg)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and then invoke local resource plugin methods. It must pass information about
+  the additional resources to the container runtime. It must detect
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
+
+For a resource plugin the following components are needed:
+- Some utility library similar to
+  https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner
+  and the code in driver-registrar.
+- *Resource controller*: a central component which handles resource allocation
+  by watching and modifying ResourceClaims.
+- *Resource node plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
```
I want to make sure I understand the proposed flow here. Let me see if I can put it into my own word. Please correct anything I got wrong.

First -- it sounds like plugins no longer "pre-advertise" their resources in the same way anymore. I.e. the api-server / scheduler no longer has a "count" of how many of a given resource a plugin is in charge of. Instead, the plugins now advertise their resources via a custom CRD, so users can still see what's available and request them in their pod spec (I assume with some sort of accounting of how many of each type are available still somewhere).

With this in place, a container within a pod spec can now "declare" what resources it would like to get access to, and a set of `ResourceClaim`s will get generated on its behalf.  Plugins on each node are then responsible for monitoring these `ResourceClaim`s and if a plugin can satisfy it, it declares that it can by adding its node's name to the `PotentialNodes` list in the `ResourceClaimStatus` field of the `ResourceClaim` (essentially locking these resources on each node until the scheduler has made a decision on which one to use to satisfy the claim).

Once this is done, the scheduler plugin kicks in to figure out which node is best suited to satisfy all of the the resource requests across all containers of the pod and eventually schedules the pod there (unlocking the resources on all nodes except the one where the pod is being scheduled).

After being scheduled, the kublet then does whatever is necessary to actually run the pod and grant it the requested resources. It does this by coordinating with the plugin to figure out what fields need to be added to the CRI / CDI interfaces to satisfy the request.

pohly committed:
Merge pull request #6 from pohly/summary-update

summary update

pohly committed:
Merge remote-tracking branch 'pohly/cdi-parameters' into dynamic-resource-allocation

sftim commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+changes to the scheduler.
+
+Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
```
Could this API also support managing resources that are cluster- or namespace- specific and _not_ attached to a specific node?
For example:
- an NFS filesystem
- a NAT gateway for the cluster network
- a SAN fabric

I'm not calling for that now. I'm asking whether folks think this proposed design leaves room for that kind of extensibility.

pohly replied:
> Could this API also support managing resources that are cluster- or namespace- specific and not attached to a specific node?

Yes, network-attached resources are supported. But the expectation is that there's still a node-local part of the plugin which then makes those network-attached resources locally available when a pod needs them. kubelet can't do it itself without that part because kubelet has no knowledge what the resource is.

sftim replied:
Imagine I have a userland NFS client and for some of my Pods I don't want the node involved at all, I just want to connect using TCP. (NetworkPolicy etc or my CNI plugin can take care of packet filtering etc). Is that feasible?

eero-t replied:
Userland NFS client inside the Pod, nothing involving the host (node)?

I do not understand what specific problem you want Kubernetes to solve for you in this case. If it's not about node resource allocation/access/management, and pod scheduling to nodes based on that, why involve Kubernetes resource allocation at all?

sftim replied:
Kubernetes is a really useful API for managing cloud things. Yes, it started out with containers but there same architecture is good for more than that.
I'm asking if this dynamic resource allocation design leaves room for resources that you can manage, claim, have a ResourceQuota for - but the thing using that resource is _not_ a Pod.

pohly replied:
You could create a stand-alone `ResourceClaim` with immediate allocation, then use it "somehow" once it is allocated. But whoever consumes it then may need to know what information the plugin puts into the `ResourceClaim.Status`. The API would support that. In this case, the node-local plugin wouldn't be necessary.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+the new API is going to be implemented independently of the existing device
+manager, there's little risk of breaking stable APIs.
+
+The new API is inspired by the volume provisioning API and uses similar
```
I'd change this first line to include the word 'existing', i.e.:
> The new API is inspired by the existing volume provisioning API ...

Also, is there. KEP for the volume provisioning API? It would be good to reference it here, if so.

pohly replied:
The best I know of is https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md#provisioning-and-deleting which I'll link to.

There's no proper KEP for it.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+different vendors can be installed at the same time without making further
+changes to the scheduler.
+
+Communication with kubelet is
```
s/with/with the/

pohly replied:
Ack.

I've seen it used like that elsewhere, but never quite got used to it because it goes against my understanding of the definite article in English: I see "kubelet" as a name, and those are used without it.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+
+Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
+mechanism, using a new plugin type and a new gRPC interface. Towards
```
The use of "towards" here sounds awkward. Consider rephrasing.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+countable per-node extended resources was added. The device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources are too limited. This KEP addresses
```
s/are/is

(I realize the subject doesn't match the verb with this change, but for some reason this sounds better in english with `is`).

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+the following limitations:
+
+- *Device initialization*: When starting a workload I’d like to have
+  the device reconfigured or reprogrammed by the orchestration. For
```
s/by the orchestration/at the orchestration layer/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+countable per-node extended resources was added. The device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources are too limited. This KEP addresses
+the following limitations:
```
I'd probably reword "This KEP addresses the following limitations:" to "This KEP addresses limitations of the current approach for the following use cases:"

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device because parameters are too limited
```
s/allow/allow one/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying workload I’d like to specify
```
s/workload/a workload/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
```
s/the device/a device/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
```
s/would/will/
s/workload/the workload/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
+  the node without a device. GPU and crypto-offload engines are
```
s/the node/a node/

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+and make progress.
+-->
+
+* An extension of the model that kube-scheduler has about
```
These should all begin with a verb, just as all of the "Goals" bullet points do.
Also, can we put newlines between each bullet here.

pohly replied:
> These should all begin with a verb, just as all of the "Goals" bullet points do

So "More flexibility" and "User-friendly API" are considered verbs? Sorry, couldn't resist. :sweat_smile: 

Seriously, I am wondering whether I should also update those. I'll leave them for now and only change the non-goals because there verbs indeed work better.

klueska commented on keps/sig-node/xxx-dynamic-resource-allocation/README.md:
```diff
@@ -0,0 +1,1428 @@
+
+![components](./components.svg)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and then invoke local resource plugin methods. It must pass information about
+  the additional resources to the container runtime. It must detect
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
```
Can we duplicate this in the intro as well? I think it will be useful to see this list very early on before reading further. We can probably duplicate this list verbatim and then have a sentence following it that forward references this section and says we will go into details on each of these here.

pohly replied:
Done.

klueska reviewed (commented):
Just did a pass over the intro --> beginning of the implementation section.

Most of my comments are superficial, but make for smoother reading. I don't think it's worth nitpicking the exact design details at this stage -- it's more important that we motivate the work properly and get people wanting to dig into the KEP and give feedback. As such, I think once these changes are integrated we can confidently put this in front of more people.

Other than that, I'm still not clear on exactly how we will handle topology alignment with this new design (though I know @kad has strong opinions about this), but I think that discussion (and possible addition to the KEP) can wait until we have this in front of more people.

Thanks for driving this @pohly.

pohly committed:
review feedback

https://github.com/pohly/enhancements/pull/1#pullrequestreview-817516213

pohly committed:
Merge pull request #7 from pohly/intro-wording

review feedback

pohly commented:
Moved to https://github.com/kubernetes/kubernetes/pull/111023
